promql/engine.go

```diff
@@ -1104,58 +1099,71 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 		enh := &EvalNodeHelper{Out: make(Vector, 0, 1)}
 		// Process all the calls for one time series at a time.
 		it := storage.NewBuffer(selRange)
-		for i, s := range selVS.Series {
-			ev.currentSamples -= len(points)
-			points = points[:0]
-			it.Reset(s.Iterator())
-			ss := Series{
-				// For all range vector functions, the only change to the
-				// output labels is dropping the metric name so just do
-				// it once here.
-				Metric: dropMetricName(selVS.Series[i].Labels()),
-				Points: getPointSlice(numSteps),
-			}
-			inMatrix[0].Metric = selVS.Series[i].Labels()
-			for ts, step := ev.startTimestamp, -1; ts <= ev.endTimestamp; ts += ev.interval {
-				step++
-				// Set the non-matrix arguments.
-				// They are scalar, so it is safe to use the step number
-				// when looking up the argument, as there will be no gaps.
-				for j := range e.Args {
-					if j != matrixArgIndex {
-						otherInArgs[j][0].V = otherArgs[j][0].Points[step].V
-					}
-				}
-				maxt := ts - offset
-				mint := maxt - selRange
-				// Evaluate the matrix selector for this series for this step.
-				points = ev.matrixIterSlice(it, mint, maxt, points)
-				if len(points) == 0 {
-					continue
+		if selVS.Series == nil {
```

Comment:
Here's your problem I suspect, this won't work with subqueries as Series will not be nil. I'd suggest merging UnexpandedSeriesSet and Series into one field, which you have to iterate over.